src/sdk/reencrypt.ts

```diff
@@ -17,14 +18,35 @@ export const reencryptRequest =
     userAddress: string,
   ) => {
     if (!gatewayUrl) throw new Error('You must provide a reencryption URL.');
-    const data = {
-      publicKey,
-      handle,
+
+    const payload = {
       signature,
-      contractAddress,
-      userAddress,
+      verification_key: userAddress,
```

Comment:
it is the last 20 bytes of the hash of the public key, this is way more easily accessible on Ethereum than the public key. Why would you "need" the public key? Furthermore, later for the delegated reencryption, the public key might not exist (but the address will always exist, even for smart contract accounts).

Reply:
the payload is signed using the private part of this key, with the EIP712 metadata.

the signature needs to be checked at some point. right now the KMS backend does it, so it needs this public key. if the js client only sends `userAddress`, then there's no way to check the signature with the implementation we have now.

now I don't know how much the gateway should verify and how much KMS should verify. maybe the KMS doesn't need to verify this signature and we let the gateway do it (and leave it as a TODO for now).

Reply:
according to https://github.com/zama-ai/tech-spec/blob/main/kms/reencryption.md#request-a-reencryption it seems the gateway needs to collect the missing data, possibly including the real public key that corresponds to `userAddress`, and then sends everything to the KMS nodes for verification.

Reply:
sorry I still don't understand, why don't you hash the recovered public key from signature on KMS backend to check it corresponds to the address?

Reply:
right, didn't know it's possible to recover the public key from signatures. we can take a look on how that's done

Reply:
(something for the future: at the moment we disabled verification on the client side, so there's no need for this long-term public key. but if the client doesn't know its own public key, it won't be able to verify the response. this is due to how signcryption works since the intended receiver should be the party that holds the long-term public key cc @jot2re this is where the verification happens https://github.com/zama-ai/kms-core/blob/main/core/service/src/cryptography/signcryption.rs#L284)

Reply:
actually, now I'm a bit lost, if the public key is always in the signature, then the client will always have it right, since it signs the message? just not in the delegated case I imagine

Reply:
Okay, I agree that we need to implement public key recovery based on the signature.
@kc1212 More specifically in ECDSA it is possible to recover the two possible public keys that signed at message, given only the signature. Since we will have the address it is possible to figure out which of these two keys is the correct one. 
Basically this implies that we won't do typical signature validation but instead key recovery and checking if it matches an address. 

In relation to signcryption we need to modify this to not use hash-digests, but addresses instead. A client should be able to know its own address, so in that case I think the signcryption should work as well.

Reply:
However, this is not a small change so maybe we want to wait with this until we figure out if we change the reencryption to the padding approach in Q3, before stating to modify this.

Reply:
In any case an issue should be added to make sure we don't forgot about this

Reply:
Just to be sure : you are talking here about the Ethereum public key, right? Not the public key generated by libsodium for reencryption, AFAIU.
Of course the user will always control his Ethereum private key, so he could retrieve his Ethereum public key as well.
However, for UX, Metamask does NOT expose the Ethereum public key, only the address, and extracting this key is not straightforward. (For instance, for any given address of an externally owned account on Ethereum, if you want to know its public key, you must wait for it to initiate a transaction.)
It is way more straightforward to work with addresses on Ethereum, also because those not only exist for EOAs but for smart contracts as well, and smart contracts do not own a private key.

Reply:
to keep the changes to a minimum, i suggest 
1. let the gateway find the public key and then send it to the kms
2. currently the wasm client needs to be initialized using a public key, but this is not possible anymore, so we need to make this field https://github.com/zama-ai/kms-core/blob/main/core/service/src/client.rs#L111 optional (or we hack in a constant public key when wasm is initializing this client to avoid code changes in many other places, and eventually create a client that's specific for wasm since it's diverging quite a bit from the current client)

Reply:
@jatZama We only use libsodium for encryption, so yes, even for signcryption we are talking about the Ethereum public key.

Reply:
@kc1212 given that we need to change both the signature of the client and the gateway, maybe it is just as simple to change the validation to key recovery? Because I am not sure how much work it is to recover the public key in the gateway

Reply:
it should be doable following this example https://docs.rs/k256/latest/k256/ecdsa/#recovering-a-verifyingkey-from-a-signature 

we already have something that doesn't need the ethereum public key (or the kms server public key) to decrypt https://github.com/zama-ai/kms-core/blob/main/core/service/src/client.rs#L1661